tests/native/test_models.py

```diff
@@ -198,3 +227,123 @@ def test_symbolic_mixed(self):
         self.assertTrue(self.state.must_be_true(ret == 4))
+
+
+class StrcpyTest(ModelTest):
+    def _assert_concrete(self, s, d):
+        # Checks all characters are copied until the 1st that could be NULL
+        cpu = self.state.cpu
+        src = cpu.read_int(s, 8)
+        dst = cpu.read_int(d, 8)
+        offset = 0
+        while cant_be_NULL(src, self.state.constraints):
+            self.assertTrue(not issymbolic(dst))
+            self.assertEqual(src, dst)
+            offset += 1
+            src = cpu.read_int(s + offset, 8)
+            dst = cpu.read_int(d + offset, 8)
+
+        # Assert final NULL byte
+        self.assertTrue(is_definitely_NULL(src, self.state.constraints))
+        self.assertEqual(0, dst)
+        return offset
+
+    """
+    This method creates memory for a given src (with no possible NULL bytes but and a
+    final NULL byte) and dst string pointers, asserts that everything is copied from src
+    to dst until the NULL byte, and asserts the memory address returned by strcpy is
+    equal to the given dst address.
+    """
+
+    def _test_strcpy(self, string, dst_len=None):
+        # Create src and dsty strings
+        if dst_len is None:
+            dst_len = len(string)
+        cpu = self.state.cpu
+        s = self._push_string(string)
+        d = self._push_string_space(dst_len)
+        dst_vals = [None] * dst_len
+        for i in range(dst_len):
+            # Set each dst byte to a random char to simplify equal comparisons
+            c = random.randrange(1, 255)
+            cpu.write_int(d + i, c, 8)
+            dst_vals[i] = c
+
+        ret = strcpy(self.state, d, s)
+
+        # addresses should match
+        self.assertEqual(ret, d)
+        # assert everything is copied up to the 1st possible 0 is copied
+        offset = self._assert_concrete(s, d)
+
+        # Delete stack space created
+        self._pop_string_space(dst_len + len(string))
+
+    def test_concrete(self):
+        self._test_strcpy("abc\0")
+        self._test_strcpy("a\0", dst_len=10)
+        self._test_strcpy("abcdefghijklm\0")
+        self._test_strcpy("a\0", dst_len=5)
+
+    def test_concrete_empty(self):
+        self._test_strcpy("\0")
+        self._test_strcpy("\0", dst_len=10)
+
+    def test_symbolic(self):
+        BIN_PATH = os.path.join(os.path.dirname(__file__), "binaries", "sym_strcpy_test")
+        tmp_dir = tempfile.TemporaryDirectory(prefix="mcore_test_sym_")
+        m = Manticore(BIN_PATH, stdin_size=10, workspace_url=str(tmp_dir.name))
+
+        addr_of_strcpy = 0x400418
+
+        @m.hook(addr_of_strcpy)
+        def strlen_model(state):
+            state.invoke_model(strcpy)
+
+        m.run()
+        m.finalize()
+
+        # Approximate regexes for expected testcase output
+        # Example Match above each regex
+        # Manticore varies the hex output slightly per run
+        expected = [
+            # STDIN: b'\x00AAAAAAAAA'
+            "STDIN: b\\'\\\x00A.{8,32}\\'",
```

Comment:
JFYI: You should be able to use raw strings here to make these regexps more intelligible:

```python
r"STDIN: b'\xffA\x00\xff\xff\xff\xff\xff\xff\xff'"
```